### ISSUE TITLE
Figure out new color waveforms used by CDJ-3000

### ISSUE THREAD

**Issue by brunchboy** [Deep-Symmetry/dysentery#33]
There’s a new database export file, `ANLZ0000.2EX`, which has the normal tagged type structure we see in all analysis files, and it has two new tags, `PWV6` and `PWV7` which almost certainly contain waveforms in the new color scheme used by the CDJ-3000. @jan2000, any chance you’ve started figuring out how they work?

**Comment by brunchboy** [comment]
Hmm, there is also a tiny `PWVC` tag at the end of the file. Maybe that is a color table used by the other two tags?

**Comment by EvanPurkhiser** [comment]
Hoping these new formats do not have a convoluted encoding schema like the colored preview waveform did

**Comment by brunchboy** [comment]
Absolutely! The pictures I’ve seen look like there are basically just three heights for each segment, with different colors for each (I’m guessing they represent lows, mids, and highs). That seems simpler than the six-variable mess you are alluding to! But I am always ready to be horrified when we get down to the details… Also, we should now expect they might be XOR-masked, sigh.

**Comment by nudge** [comment]
Good news everyone!

Took a quick look at the `PWV6` format and indeed it looks to be split across frequency bands. In the preview format, the bands appear to be stacked on one another. There is also some funky scaling going on, not sure how the height coefficients are calculated so using some fixed ones for the time being. Needs some work here, it is especially evident in the sine chirp signal (see second set of images).

```
byte 0: mid frequency
byte 1: high frequency
byte 2: low frequency
```

This is what it looks like compared to rekordbox for a normal track, not bad:

<img width="1723" alt="Screen Shot 2021-02-15 at 10 31 11 pm" src="https://user-images.githubusercontent.com/360201/107994851-bfb99c00-6fdd-11eb-9166-4a780b214d62.png">

<img width="1401" alt="Screen Shot 2021-02-15 at 10 31 17 pm" src="https://user-images.githubusercontent.com/360201/107994854-c34d2300-6fdd-11eb-99f5-fa78c22bc9e0.png">

Using a generated sine chirp signal, it is evident some scaling correction (plus clipping?) needs to be done:

<img width="1509" alt="Screen Shot 2021-02-15 at 10 35 28 pm" src="https://user-images.githubusercontent.com/360201/107995019-2343c980-6fde-11eb-8e2e-e5810838a8ca.png">
<img width="1394" alt="Screen Shot 2021-02-15 at 10 24 38 pm" src="https://user-images.githubusercontent.com/360201/107995026-26d75080-6fde-11eb-8e48-f1c3bc0df8ae.png">

Anyway sample code is up at https://gist.github.com/nudge/eca9f07ff684955b8a9bb5944614c3d7, based on the super cool work from @jan2000!

**Comment by brunchboy** [comment]
Nice work, that was FAST!

**Comment by agressiva** [comment]
Does any one convert that code to C ansi ?

**Comment by brunchboy** [comment]
I am finally getting around to implementing this in Beat Link.

**Comment by brunchboy** [comment]
Hey, @nudge, I have this working in Beat Link for the preview half-waveforms, but I can’t figure out what is going on with the extra colors in the full-size scrolling waveform? The sample code does not address that.

**Comment by brunchboy** [comment]
Hmm, I think I have an idea. Maybe instead of stacking the colors and making them opaque, as is done in the preview, I will try drawing the mid first, opaque, then the high over that translucently, so we get mixing as seen in rekordbox, and then finally the low opaque over that.

**Comment by brunchboy** [comment]
That approach is looking much closer. I ended up having to do my own version of blending the low and mid-range colors and drawing the overlap separately to get the colors right, but it is looking much closer to what rekordbox does now. I’ll push these latest changes to `main`. The scaling is still not quite right, in fact it is becoming pretty clear that there is non-linear scaling of some sort involved for the detail waveforms, but that level of nicety will have to wait for another day (and any ideas are super welcome!) This is useful enough as a starting point.

Anyone who wants to see what it is looking like can add this to your Global Setup Expression until I get a UI added to control this:

```clojure
(.setPreferredStyle (org.deepsymmetry.beatlink.data.WaveformFinder/getInstance)
  org.deepsymmetry.beatlink.data.WaveformFinder$WaveformStyle/THREE_BAND)
```

**Comment by brunchboy** [comment]
See https://github.com/Deep-Symmetry/beat-link/commit/392442b4858e895f83e6456882f52df9aea247c0
and https://github.com/Deep-Symmetry/beat-link/commit/86184f3aef211f614a84deb7fbab3478e347bf2d

**Comment by brunchboy** [comment]
And here is an example of what it is looking like. Now I need to practice my piano, then update the dysentery documentation for what has been implemented so far.

<img width="1074" alt="Image" src="https://github.com/user-attachments/assets/d1beb768-113e-4fbb-9c7f-355d0c76c1ed" />

**Comment by evanpurkhiser** [comment]
Looks incredible!! awesome work!
